src/masternodeman.cpp

```diff
@@ -644,6 +671,416 @@ void CMasternodeMan::ProcessMessage(CNode* pfrom, std::string& strCommand, CData
+            ProcessVerifyBroadcast(pfrom, mnv);
+        }
+    }
+}
+
+void CMasternodeMan::DoFullVerificationStep()
+{
+    if(activeMasternode.vin == CTxIn()) return;
+
+    std::vector<std::pair<int, CMasternode> > vecMasternodeRanks = GetMasternodeRanks(pCurrentBlockIndex->nHeight - 1, MIN_POSE_PROTO_VERSION);
+
+    LOCK(cs);
+
+    int nCount = 0;
+    int nCountMax = std::max(10, (int)vMasternodes.size() / 100); // verify at least 10 masternode at once but at most 1% of all known masternodes
+
+    int nMyRank = -1;
+    int nRanksTotal = (int)vecMasternodeRanks.size();
+
+    // send verify requests only if we are in top MAX_POSE_RANK
+    std::vector<std::pair<int, CMasternode> >::iterator it = vecMasternodeRanks.begin();
+    while(it != vecMasternodeRanks.end()) {
+        if(it->first > MAX_POSE_RANK) {
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Must be in top %d to send verify request\n",
+                        MAX_POSE_RANK);
+            return;
+        }
+        if(it->second.vin == activeMasternode.vin) {
+            nMyRank = it->first;
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Found self at rank %d/%d, verifying up to %d masternodes\n",
+                        nMyRank, nRanksTotal, nCountMax);
+            break;
+        }
+        ++it;
+    }
+
+    // send verify requests to up to nCountMax masternodes starting from
+    // (MAX_POSE_RANK + nCountMax * (nMyRank - 1) + 1)
+    int nOffset = MAX_POSE_RANK + nCountMax * (nMyRank - 1);
+    if(nOffset >= (int)vecMasternodeRanks.size()) return;
+
+    std::vector<CMasternode*> vSortedByAddr;
+    BOOST_FOREACH(CMasternode& mn, vMasternodes) {
+        vSortedByAddr.push_back(&mn);
+    }
+
+    it = vecMasternodeRanks.begin() + nOffset;
+    while(it != vecMasternodeRanks.end()) {
+        if(it->second.IsPoSeVerified() || it->second.IsPoSeBanned()) {
+            LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Already %s%s%s masternode %s address %s, skipping...\n",
+                        it->second.IsPoSeVerified() ? "verified" : "",
+                        it->second.IsPoSeVerified() && it->second.IsPoSeBanned() ? " and " : "",
+                        it->second.IsPoSeBanned() ? "banned" : "",
+                        it->second.vin.prevout.ToStringShort(), it->second.addr.ToString());
+            ++it;
+            continue;
+        }
+        LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Verifying masternode %s rank %d/%d address %s\n",
+                    it->second.vin.prevout.ToStringShort(), it->first, nRanksTotal, it->second.addr.ToString());
+        if(SendVerifyRequest((CAddress)it->second.addr, vSortedByAddr) && ++nCount >= nCountMax) break;
+        ++it;
+    }
+
+    LogPrint("masternode", "CMasternodeMan::DoFullVerificationStep -- Sent verification requests to %d masternodes\n", nCount);
+}
+
+void CMasternodeMan::CheckSameAddr()
```

Comment:
Since all nodes run this function, it seems like this will lead to a very large number of connections to any MN with multiple IP's.  I think this will likely have the effect of knocking the real MN offline and hence is not really any better than just dropping all MN's with the same IP.